ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java

```diff
@@ -2950,21 +2957,32 @@ public Partition createPartition(Table tbl, Map<String, String> partSpec) throws
     int size = addPartitionDesc.getPartitionCount();
     List<org.apache.hadoop.hive.metastore.api.Partition> in =
         new ArrayList<org.apache.hadoop.hive.metastore.api.Partition>(size);
-    AcidUtils.TableSnapshot tableSnapshot = AcidUtils.getTableSnapshot(conf, tbl, true);
     long writeId;
     String validWriteIdList;
-    if (tableSnapshot != null && tableSnapshot.getWriteId() > 0) {
-      writeId = tableSnapshot.getWriteId();
-      validWriteIdList = tableSnapshot.getValidWriteIdList();
+
+    // In case of replication, get the writeId from the source and use valid write Id list
+    // for replication.
+    if (addPartitionDesc.getReplicationSpec().isInReplicationScope() &&
+        addPartitionDesc.getPartition(0).getWriteId() > 0) {
```

Comment:
I think, in replication flow and for transactional table, the 2nd condition should be true. If no valid writeId obtained from source, then need to fail. We cannot fall back to default logic.

Reply:
writeId will be 0 for non-transactional tables. Also this is createPartitions code, which may get executed for partitions created when writeId 0 for transactional tables as well.

Reply:
I think, this assumption of table type is non-transactional (based on writeId=0) and ignoring failure case is not right.
We can explicitly check if it is transactional table or not and then do necessary checks. If writeId comes as 0 for transactional table, then it is error flow.